plugins/pixart-rf/fu-pxi-wireless-device.c

```diff
@@ -0,0 +1,606 @@
+/*
+ * Copyright (C) 2020 Jimmy Yu <Jimmy_yu@pixart.com>
+ * Copyright (C) 2021 Richard Hughes <richard@hughsie.com>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+#include "config.h"
+
+#ifdef HAVE_HIDRAW_H
+#include <linux/hidraw.h>
+#include <linux/input.h>
+#endif
+
+#include <fwupdplugin.h>
+
+#include "fu-pxi-wireless-device.h"
+#include "fu-pxi-firmware.h"
+#include "fu-pxi-receiver-device.h"
+#include "fu-pxi-common.h"
+
+struct _FuPxiWirelessDevice {
+	FuDevice		 parent_instance;
+	struct ota_fw_state	 fwstate;
+	guint8			 sn;
+	struct ota_fw_dev_model	 model;
+};
+
+G_DEFINE_TYPE (FuPxiWirelessDevice, fu_pxi_wireless_device, FU_TYPE_DEVICE)
+
+static void
+fu_pxi_wireless_device_to_string (FuDevice *device, guint idt, GString *str)
+{
+	FuPxiWirelessDevice *self = FU_PXI_WIRELESS_DEVICE (device);
+	fu_pxi_ota_fw_state_to_string (&self->fwstate, idt, str);
+	fu_common_string_append_kv (str, idt, "ModelName", (gchar*)self->model.name);
+	fu_common_string_append_kx (str, idt, "ModelType", self->model.type);
+	fu_common_string_append_kx (str, idt, "ModelTarget", self->model.target);
+}
+
+static gboolean
+fu_pxi_wireless_device_set_feature (FuDevice *self, const guint8 *buf, guint bufsz, GError **error)
+{
+#ifdef HAVE_HIDRAW_H
+	if (g_getenv ("FWUPD_PIXART_RF_VERBOSE") != NULL)
+		fu_common_dump_raw (G_LOG_DOMAIN, "SetFeature", buf, bufsz);
+	return fu_udev_device_ioctl (FU_UDEV_DEVICE (self),
+				     HIDIOCSFEATURE(bufsz), (guint8 *) buf,
+				     NULL, error);
+#else
+	g_set_error_literal (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_NOT_SUPPORTED,
+			     "<linux/hidraw.h> not available");
+	return FALSE;
+#endif
+}
+
+static gboolean
+fu_pxi_wireless_device_get_feature (FuDevice *self, guint8 *buf, guint bufsz, GError **error)
+{
+#ifdef HAVE_HIDRAW_H
+	if (!fu_udev_device_ioctl (FU_UDEV_DEVICE (self),
+				   HIDIOCGFEATURE(bufsz), buf,
+				   NULL, error)) {
+		return FALSE;
+	}
+	if (g_getenv ("FWUPD_PIXART_RF_VERBOSE") != NULL)
+		fu_common_dump_raw (G_LOG_DOMAIN, "GetFeature", buf, bufsz);
+	return TRUE;
+#else
+	g_set_error_literal (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_NOT_SUPPORTED,
+			     "<linux/hidraw.h> not available");
+	return FALSE;
+#endif
+}
+
+static FuPxiReceiverDevice *
+fu_pxi_wireless_device_get_parent (FuDevice *self, GError **error)
+{
+	FuDevice *parent = fu_device_get_parent (FU_DEVICE (self));
+	if (parent == NULL) {
+		g_set_error_literal (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_INTERNAL,
+				     "no parent set");
+		return NULL;
+	}
+	return FU_PXI_RECEIVER_DEVICE (FU_UDEV_DEVICE (parent));
+}
+
+static gboolean
+fu_pxi_wireless_device_open (FuDevice *device, GError **error)
+{
+	FuDevice *parent = fu_device_get_parent (device);
+	if (parent == NULL) {
+		g_set_error_literal (error,
```

Comment:
maybe this should be `fu_pxi_wireless_device_get_parent`

Reply:
Yes, I will fix it.

Reply:
@jyu99 can you push fixes to this branch please.

Reply:
After I check it again. I think it's fine to use `fu_device_get_parent`. Could I keep using this?

Reply:
My point is there is code duplication here, with exact same thing in `fu_pxi_wireless_device_get_parent`.